Remove RenderBuilding

reaperrr opened the pull request:
Replaces the four traits inheriting `RenderBuilding` with solutions based on `WithSpriteBody`.
Replaces `RenderBuilding` with `WithSpriteBody` + `RenderSprites` + `AutoSelectionSize`.

Supercedes #8753.

reaperrr committed:
RenderBuildingCharge -> WithChargeAnimation

reaperrr committed:
RenderBuildingSilo -> WithSiloAnimation

reaperrr committed:
RenderBuildingTurreted -> WithTurretedSpriteBody

reaperrr committed:
RenderBuildingWall -> WithWallSpriteBody

reaperrr committed:
Remove RenderBuilding

reaperrr committed:
Extend WithSpriteBody funtionality

Move building placement range circle to PlaceBuilding, add
PauseAnimationWhenDisabled.

Mailaender commented:
This breaks all make animations for buildings. Tested the cnc mod.

reaperrr commented:
Fixed in #8753, I'll update this once #8753 is merged.

reaperrr commented:
This now supercedes #8753, since I've had major trouble handling the issues that splitting this in two PRs caused.

Mailaender commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -56,7 +56,8 @@ public override Activity Tick(Actor self)
 					continue;
 
-				hostBuilding.Trait<RenderBuilding>().PlayCustomAnim(hostBuilding, "active");
+				if (hostBuilding.HasTrait<WithSpriteBody>())
```
This is an unnecessary trait lookup. You are already doing a null check in line 53.

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -159,7 +159,7 @@ public Building(ActorInitializer init, BuildingInfo info)
 		public void Created(Actor self)
 		{
-			if (SkipMakeAnimation || !self.HasTrait<WithMakeAnimation>())
+			if (SkipMakeAnimation || (!self.HasTrait<WithMakeAnimation>() && !self.HasTrait<WithMakeAnimation>()))
```
This is duplicating the check, probably an auto-rename left over. You should think about squashing things and getting rid of the workaround commits.

Mailaender commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -25,6 +25,9 @@ public class WithSpriteBodyInfo : UpgradableTraitInfo, IRenderActorPreviewSprite
 		[Desc("Animation to play when the actor is idle.")]
 		[SequenceReference] public readonly string Sequence = "idle";
 
+		[Desc("Pause animation when actor is disabled (examples: low on power, EMP).")]
```
Try avoiding mod specific language in general trait documentation.

Mailaender commented on OpenRA.Mods.Common/Traits/SupportPowers/GrantUpgradePower.cs:
```diff
@@ -30,6 +30,9 @@ class GrantUpgradePowerInfo : SupportPowerInfo
 		public readonly string GrantUpgradeSound = "ironcur9.aud";
 
+		[Desc("Sequence to play for granting actor when activated.")]
+		[SequenceReference]public readonly string GrantUpgradeSequence = "active";
```
Missing space (very minor issue though).

reaperrr committed:
Fixed production overlay ZOffsets

reaperrr commented:
Updated.

Mailaender commented:
The defense buildings probably need some manual selection size tweaking:

![image](https://cloud.githubusercontent.com/assets/756669/9566369/2d97a8a4-4f03-11e5-8b1d-9071f79a75fc.png)

Mailaender commented:
GDI upgrade center offset is wrong for both place building and plugins:

![image](https://cloud.githubusercontent.com/assets/756669/9566406/0ae51312-4f05-11e5-8771-ddc99022bf2a.png)

Mailaender commented:
Crash when firing the Tiberian Sun Nod nuke/cluster/whatever missile:

```
Exception of type `System.InvalidOperationException`: Actor does not have trait of type `OpenRA.Mods.Common.Traits.RenderSimple`
  at OpenRA.TraitDictionary+TraitContainer`1[OpenRA.Mods.Common.Traits.RenderSimple].Get (UInt32 actor) [0x0003e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/TraitDictionary.cs:150 
  at OpenRA.TraitDictionary.Get[RenderSimple] (OpenRA.Actor actor) [0x00013] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/TraitDictionary.cs:96 
  at OpenRA.Actor.Trait[RenderSimple] () [0x0000d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:212 
  at OpenRA.Mods.Common.Traits.NukePower.Activate (OpenRA.Actor self, OpenRA.Order order, OpenRA.Mods.Common.Traits.SupportPowerManager manager) [0x00075] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs:74 
  at OpenRA.Mods.Common.Traits.SupportPowerInstance.Activate (OpenRA.Order order) [0x00053] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs:238 
  at OpenRA.Mods.Common.Traits.SupportPowerManager.ResolveOrder (OpenRA.Actor self, OpenRA.Order order) [0x00029] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs:108 
  at OpenRA.Network.UnitOrders.ProcessOrder (OpenRA.Network.OrderManager orderManager, OpenRA.World world, Int32 clientId, OpenRA.Order order) [0x00c87] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Network/UnitOrders.cs:305 
  at OpenRA.Network.OrderManager.Tick () [0x000a1] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Network/OrderManager.cs:177 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x0019d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:463 
  at OpenRA.Game.LogicTick () [0x0004e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:495 
  at OpenRA.Game.Loop () [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:619 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:659 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:117 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:39 
AL lib: (EE) alc_cleanup: 1 device not closed
```

reaperrr commented:
The upgrade center offset issue is known and on bleed already, but I might as well fix it now.

reaperrr committed:
Fix d2k NukePower sequence requirement

reaperrr commented:
Updated.
- Fixed `NukePower` crash and unhardcoded its activation sequence (squashed into 9da56f5).
- Allowed to disable `NukePower` activation anim via null or empty.
- Fixed that construction yards with `WithBuildingPlacedAnimation` would not resume their idle animation after a building was placed (squashed into 9da56f5).
- Fixed that GDI Comp. Tower in TS had both a wall body and a normal sprite body (squashed into 9da56f5).
- Fixed various overlay ZOffsets.

reaperrr commented:
Fixed the TD selection boxes and improved TS offsets in #9204, as these are not caused by this PR.

Mailaender commented:
Can't spot regressions anymore. :white_check_mark:

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -63,6 +66,16 @@ public string NormalizeSequence(Actor self, string sequence)
+		{
+			DefaultAnimation.PlayRepeating(NormalizeSequence(self, Info.Sequence));
+
+			if (Info.PauseAnimationWhenDisabled)
```
Please move this `if` (and its contents) to the constructor so only the `PlayRepeating()` above remains in this method.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:
```diff
@@ -105,7 +105,8 @@ public void Tick(Actor self)
 			if (dockedHarv != null && dockedHarv.IsDead)
 			{
-				self.Trait<RenderBuilding>().CancelCustomAnim(self);
+				var wsb = self.Trait<WithSpriteBody>();
```
Please cache this in the constructor.

penev92 commented on mods/ts/sequences/structures.yaml:
```diff
@@ -1163,7 +1170,7 @@ namisl:
 		Length: 18
 		ShadowStart: 18
-	active: ntmisl_a
+	active: ntmisl_a # TODO: This is an overlay
```
How about "TODO: This is unused! (Because ...)"

reaperrr replied:
No, because it's not unused. It's just treated like an animation rather than an overlay at the moment. I can try to make the comment clearer on that, though.

penev92 commented:
:+1:  with the 3 comments above.

reaperrr committed:
Cache WithSpriteBody in Refinery constructor

reaperrr committed:
Clarify TODO on TS missile silo sequences

reaperrr committed:
Move WithSpriteBody pause check to constructor

reaperrr commented:
Updated.

penev92 commented:
Thanks! :+1: 

You can now file a PR to finally remove `RenderSimple`. ;)

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/2efcf22959d4fa2bbec1de835aece1b6c4fbb94c...7eb0241836c22689d1fb82ef1a9fc14a70677348)

reaperrr commented:
> You can now file a PR to finally remove RenderSimple. ;)

That's ready and waiting to be pushed. ;)
